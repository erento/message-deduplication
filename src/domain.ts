export enum DeliveryInfoState {
    Delivered,
    InProgress,
}

export interface DeliveryInfo {
    state: DeliveryInfoState;
    createdTime: string;
}

export interface DeliveryStorage<T> {
    get (key: string): Promise<T | undefined>;

    set (key: string, value: T, duration?: number): Promise<void>;
}
