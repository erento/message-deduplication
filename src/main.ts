import * as moment from 'moment';
import {DeliveryInfo, DeliveryInfoState, DeliveryStorage} from './domain';
import {getDeliveryStorage} from './storage';

const storage: DeliveryStorage<DeliveryInfo> = getDeliveryStorage(
    process.env.MD_IN_MEMORY_ONLY ? !!process.env.MD_IN_MEMORY_ONLY : true,
);

const MAX_ACKNOWLEDGE_TIME: { unit: moment.DurationInputArg2, value: moment.DurationInputArg1 } = {
    unit: 'seconds',
    value: process.env.MD_MAX_ACKNOWLEDGE_TIME || 900,
};

export async function setInProgress (messageId: string, subscriberName: string): Promise<void> {
    const messageDeliveryKey: string = getMessageDeliveryKey(messageId, subscriberName);
    await storage.set(messageDeliveryKey, {
        state: DeliveryInfoState.InProgress,
        createdTime: moment().utc().toISOString(),
    });
}

export async function setAsDelivered (messageId: string, subscriberName: string): Promise<void> {
    const messageDeliveryKey: string = getMessageDeliveryKey(messageId, subscriberName);

    const message: DeliveryInfo | undefined = await storage.get(messageDeliveryKey);
    if (!message) {
        throw new Error(`Couldn't find a message.`);
    }

    if (message.state !== DeliveryInfoState.InProgress) {
        throw new Error(
            `The message is not in the correct state. Message should be in progress to be marked as delivered. ` +
            `Current message: "${JSON.stringify(message)}".`,
        );
    }

    await storage.set(messageDeliveryKey, {
        state: DeliveryInfoState.InProgress,
        createdTime: moment().utc().toISOString(),
    });
}

export async function canStartProcessing (messageId: string, subscriberName: string): Promise<boolean> {
    if (!messageId || !subscriberName) {
        throw new Error(`Message ID "${messageId}" and subscriber name "${subscriberName}" have to be provided.`);
    }

    const messageDeliveryKey: string = getMessageDeliveryKey(messageId, subscriberName);

    const deliveryInfo: DeliveryInfo | undefined = await storage.get(messageDeliveryKey);
    if (deliveryInfo !== undefined) {
        if (deliveryInfo.state === DeliveryInfoState.Delivered) {
            return false;
        }

        const expiredTime: moment.Moment = moment().utc().subtract(MAX_ACKNOWLEDGE_TIME.value, MAX_ACKNOWLEDGE_TIME.unit);
        if (deliveryInfo.state === DeliveryInfoState.InProgress) {
            if (moment(deliveryInfo.createdTime).utc().isAfter(expiredTime)) {
                return false;
            }
        }
    }

    return true;
}

function getMessageDeliveryKey (messageId: string, subscriberName: string): string {
    return `${messageId || ''}_${subscriberName || ''}`;
}
