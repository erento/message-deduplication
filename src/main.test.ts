import * as moment from 'moment';

const mockGetter: jest.Mock = jest.fn();
const mockSetter: jest.Mock = jest.fn();

jest.mock('./storage', () => {
    return {
        getDeliveryStorage: (): DeliveryStorage<DeliveryInfo> => {
            return {
                _type: 'in-memory',
                get: mockGetter,
                set: mockSetter,
            };
        },
    };
});

import {CanDeliver, DeliveryInfo, DeliveryInfoState, DeliveryStorage} from './domain';
import {canBeDelivered, setAsDelivered, setInProgress} from './main';

const messageId: string = '121212121';
const subscriberName: string = 'some service_random-name2!';
const expectedKey: string = '121212121_some service_random-name2!';

describe('message deduplication', (): void => {
    beforeEach((): void => {
        mockGetter.mockReset();
        mockSetter.mockReset();
    });

    describe('set message', (): void => {
        test('should set a message to in progress', async (): Promise<void> => {
            await expect(setInProgress(messageId, subscriberName)).resolves.toBeUndefined();
            expect(mockSetter).toHaveBeenCalledWith(expectedKey, <DeliveryInfo> {
                state: DeliveryInfoState.InProgress,
                createdTime: expect.stringContaining(moment().utc().toISOString().substr(0, 17)),
            });
        });
        test('should set a message to delivered', async (): Promise<void> => {
            mockGetter.mockResolvedValue(<DeliveryInfo> {
                state: DeliveryInfoState.InProgress,
                createdTime: moment().utc().toISOString(),
            });
            await expect(setAsDelivered(messageId, subscriberName)).resolves.toBeUndefined();
            expect(mockSetter).toHaveBeenCalledWith(expectedKey, <DeliveryInfo> {
                state: DeliveryInfoState.Delivered,
                createdTime: expect.stringContaining(moment().utc().toISOString().substr(0, 17)),
            });
        });

        test('should fail to set a message to delivered if message doesn\'t exist', async (): Promise<void> => {
            mockGetter.mockResolvedValue(undefined);
            await expect(setAsDelivered(messageId, subscriberName)).rejects.toThrowErrorMatchingSnapshot();
            expect(mockSetter).not.toHaveBeenCalled();
        });

        test('should fail to set a message to delivered if is not in progress', async (): Promise<void> => {
            mockGetter.mockResolvedValue(<DeliveryInfo> {
                state: DeliveryInfoState.Delivered,
                createdTime: moment().toISOString(),
            });
            await expect(setAsDelivered(messageId, subscriberName)).rejects.toThrowError();
            expect(mockSetter).not.toHaveBeenCalled();
        });
    });

    describe('if message can be processed', (): void => {
        test('should throw an error in case message ID or subscriber name is not provided', async (): Promise<void> => {
            await expect(canBeDelivered(<any> null, <any> null)).rejects.toThrowErrorMatchingSnapshot();
            await expect(canBeDelivered(<any> undefined, <any> undefined)).rejects.toThrowErrorMatchingSnapshot();
            await expect(canBeDelivered('', '')).rejects.toThrowErrorMatchingSnapshot();
            await expect(canBeDelivered(<any> NaN, <any> NaN)).rejects.toThrowErrorMatchingSnapshot();
            await expect(canBeDelivered('', 'a')).rejects.toThrowErrorMatchingSnapshot();
            await expect(canBeDelivered('a', '')).rejects.toThrowErrorMatchingSnapshot();
        });

        test('should return true if it is a new message', async (): Promise<void> => {
            await expect(canBeDelivered(messageId, subscriberName)).resolves.toEqual(CanDeliver.Yes);
        });

        test('should return false if it is delivered', async (): Promise<void> => {
            mockGetter.mockResolvedValue(<DeliveryInfo> {state: DeliveryInfoState.Delivered, createdTime: 'some time'});
            await expect(canBeDelivered(messageId, subscriberName)).resolves.toEqual(CanDeliver.NoAlreadyDelivered);
            expect(mockGetter).toHaveBeenCalledWith(expectedKey);
        });

        test('should return false if it is in progress and not expired', async (): Promise<void> => {
            mockGetter.mockResolvedValue(<DeliveryInfo> {
                state: DeliveryInfoState.InProgress,
                createdTime: moment().subtract(2, 'minutes').toISOString(),
            });
            await expect(canBeDelivered(messageId, subscriberName)).resolves.toEqual(CanDeliver.NoInProgress);
            expect(mockGetter).toHaveBeenCalledWith(expectedKey);
        });

        test('should return true if it is in progress but expired', async (): Promise<void> => {
            mockGetter.mockResolvedValue(<DeliveryInfo> {
                state: DeliveryInfoState.InProgress,
                createdTime: moment().utc().subtract(20, 'minutes').toISOString(),
            });
            await expect(canBeDelivered(messageId, subscriberName)).resolves.toEqual(CanDeliver.Yes);
            expect(mockGetter).toHaveBeenCalledWith(expectedKey);
        });

        test('should return true if the storage return null', async (): Promise<void> => {
            mockGetter.mockResolvedValue(null);
            await expect(canBeDelivered(messageId, subscriberName)).resolves.toEqual(CanDeliver.Yes);
            expect(mockGetter).toHaveBeenCalledWith(expectedKey);
        });
    });
});
