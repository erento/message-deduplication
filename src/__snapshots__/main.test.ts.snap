// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message deduplication if message can be processed should throw an error in case message ID or subscriber name is not provided 1`] = `"Message ID \\"null\\" and subscriber name \\"null\\" have to be provided."`;

exports[`message deduplication if message can be processed should throw an error in case message ID or subscriber name is not provided 2`] = `"Message ID \\"undefined\\" and subscriber name \\"undefined\\" have to be provided."`;

exports[`message deduplication if message can be processed should throw an error in case message ID or subscriber name is not provided 3`] = `"Message ID \\"\\" and subscriber name \\"\\" have to be provided."`;

exports[`message deduplication if message can be processed should throw an error in case message ID or subscriber name is not provided 4`] = `"Message ID \\"NaN\\" and subscriber name \\"NaN\\" have to be provided."`;

exports[`message deduplication if message can be processed should throw an error in case message ID or subscriber name is not provided 5`] = `"Message ID \\"\\" and subscriber name \\"a\\" have to be provided."`;

exports[`message deduplication if message can be processed should throw an error in case message ID or subscriber name is not provided 6`] = `"Message ID \\"a\\" and subscriber name \\"\\" have to be provided."`;

exports[`message deduplication set message should fail to set a message to delivered if message doesn't exist 1`] = `"Couldn't find a message."`;
